pkg/client/bucket/bucket.go

```diff
@@ -0,0 +1,210 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+package bucket
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/concurrency"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/rest"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+const endpoint = "platform/storage/management/v1/bucket-definitions"
+
+type (
+	// Response holds all necessary information to create and update Grail Buckets
+	Response struct {
+		BucketName string `json:"bucketName"`
+		Status     string `json:"status"`
+		Version    int    `json:"version"`
+		Data       []byte `json:"-"`
+	}
+
+	// Client abstracts the API access for Grail Buckets
+	Client struct {
+		url     string
+		client  *rest.Client
+		limiter *concurrency.Limiter
+	}
+)
+
+// NewClient creates a new client to interact with the Grail Bucket API
+func NewClient(url string, client *rest.Client) *Client {
+	return &Client{
+		url:     url,
+		client:  client,
+		limiter: concurrency.NewLimiter(5),
+	}
+}
+
+// Upsert create or updates a given Grail Bucket
+func (c Client) Upsert(ctx context.Context, id string, data []byte) (result Response, err error) {
+	if id == "" {
+		return Response{}, fmt.Errorf("id must be non empty")
+	}
+	c.limiter.ExecuteBlocking(func() {
+		result, err = c.upsert(ctx, id, data)
+	})
+	return
+}
+
+// upsert first tries to create the bucket. If it does not work, we try to update it.
+// This has the advantage that we don't need to try to fetch the object first.
+func (c Client) upsert(ctx context.Context, id string, data []byte) (Response, error) {
+	r, err := c.create(ctx, id, data)
+	if err == nil {
+		log.WithCtxFields(ctx).Debug("created object with ID %q", id)
+		return r, nil
+	}
+
+	log.WithCtxFields(ctx).Debug("failed to create new object with ID %q; trying to update existing: %w", id, err)
+
+	b, err := c.get(ctx, id)
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to get object with ID %q: %w", id, err)
+	}
+
+	r, err = c.update(ctx, b, data)
+	log.WithCtxFields(ctx).Debug("updated object with ID %q", id)
+	return r, err
+}
+
+func (c Client) create(ctx context.Context, id string, data []byte) (Response, error) {
+	u, err := url.JoinPath(c.url, endpoint)
+	if err != nil {
+		return Response{}, fmt.Errorf("faild to create sound url: %w", err)
+	}
+
+	err = setBucketName(id, &data)
+	if err != nil {
+		return Response{}, err
+	}
+
+	r, err := c.client.Post(ctx, u, data)
+	if err != nil {
+		return Response{}, fmt.Errorf("unable to create object with ID %q: %w", id, err)
```

Comment:
```suggestion
		return Response{}, fmt.Errorf("failed to create object with ID %q: %w", id, err)
```

IMO wording should be same for both client and HTTP statuscode errors